pkg/nimble/autoadv/include/nimble_autoadv.h

```diff
@@ -0,0 +1,117 @@
+ *                  - No expiration (BLE_HS_FOREVER)
+ *                  - No name
+ *
+ * @{
+ *
+ * @file
+ *
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ */
+
+#ifndef NIMBLE_AUTOADV_H
+#define NIMBLE_AUTOADV_H
+
+#include "host/ble_gap.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* @brief    Name of the device for the advertising procedure. If this is not
+ *          defined, it will be defined as NULL, resulting in not configuring
+ *          a name at all.
+*/
+#ifndef NIMBLE_AUTOADV_DEVICE_NAME
+    #define NIMBLE_AUTOADV_DEVICE_NAME NULL
+#endif
+
+/**
+* @brief    If an application is calling functions from nimble, e.g.
+ *          ble_svc_gap_device_name_set(), NIMBLE_AUTOADV_START_MANUALLY should
+ *          be set to 1 and then the application should call
+ *          nimble_autoadv_start() after all of its nimble calls to prevent
+ *          errors like BLE_HS_EBUSY.
+ *
+ *          Defined as 0 by default.
+*/
+#ifndef NIMBLE_AUTOADV_START_MANUALLY
+    #define NIMBLE_AUTOADV_START_MANUALLY 0
+#endif
+
+/**
+* @brief    Initialize autoadv module.
+*/
+void nimble_autoadv_init(void);
+
+/**
+ * @brief   Set struct for additional arguments specifying the particulars of
+ *          the advertising procedure. Uses memcpy internally.
+ *
+ * @param[in]  params   struct with customized additional arguments
+ */
+void nimble_autoadv_set_ble_gap_adv_params(struct ble_gap_adv_params *params);
+
+/**
+ * @brief   Add a new field to the given advertising data.
+ *
+ * @param[in] type      field type to add
+ * @param[in] data      payload for the field
+ * @param[in] data_len  length of the payload in bytes
+ *
+ * @return  BLUETIL_AD_OK if the new field was added
+ * @return  BLUETIL_AD_NOMEM if there is not enough space to write add field
+ */
+int nimble_autoadv_add_field(uint8_t type, const void *data, size_t data_len);
+
+/**
+ * @brief   Set the duration for the advertising procedure.
+ *
+ * @param[in]  duration_ms  duration of advertising procedure in ms
+ */
+void nimble_auto_adv_set_adv_duration(int32_t duration_ms);
+
+/**
+ * @brief   Set the callback for gap events. Callback is used for the logic when
+ *          to start the advertising procedure.
+ *
+ * @param[in] cb        The callback to associate with this advertising
+ *                      procedure. If advertising ends, the event is reported
+ *                      through this callback. If advertising results in a
+ *                      connection, the connection inherits this callback as its
+ *                      event-reporting mechanism.
+ *
+ * @param[in] cb_arg    The optional argument to pass to the callback function.
+ */
+void nimble_auto_adv_set_gap_cb(ble_gap_event_fn *cb, void *cb_arg);
+
+/**
+ * @brief   Start the automated advertising procedure.
+ *
+ *          Needs to be called manually when NIMBLE_AUTOADV_START_MANUALLY was
+ *          set to true.
+ */
+void nimble_autoadv_start(void);
```

Comment:
I forgot to make one comment here: it would also be beneficial to add one additional function to this module: `nimble_autoadv_stop()`. With this, one can pause advertising e.g. while changing parameters and advertising data...

Reply:
Cool, I pushed a proposal. A new boolean variable is introduced to reflect the paused state. Checking for `ble_gap_adv_active()` would not catch this case.

Reply:
Aaahh now I understand what you meant. With your suggestion the user would be able to set the data while advertising and then he has to call stop and start manually. But this looks dirty to me, as `nimble_autoadv_set_ble_gap_adv_params` modifies `_advp` which is given by reference to `ble_gap_adv_start`. I don't now about the internal workflow of `ble_gap_adv_start`, but I guess that could change things while advertising?

Reply:
Not quite. I see the following two situations in which the update functions are used:
1. advertising is active while they are called: in this case, I would expect the new parameters to be applied and advertising is continued with the updated parameters. meaning that in practice advertising is stopped for a brief moment, params are updated and advertising is continued
2. advertising is not active while they are called: in this case merely the local state is updated and nothing else happens

Importantly: both cases are complete independent of the value of `NIMBLE_AUTOADV_START_MANUALLY`. They only depend on the advertising state at the moment when the update functions are called, so the value of `ble_gap_adv_active()` is all we need to care about here...

Reply:
Ok now I see it. I pushed yet another fixup commit ;)